HDDS-8041. Ozone FS Client should fail to connect if wrong hostname/serviceID in URI.

sadanand48 opened the pull request:
## What changes were proposed in this pull request?
On providing an incorrect hostname/service ID in ofs URI , the filesystem client  instead of failing , retries till exhaustion. Also the default retry config for client retries is too high currently  (500). Considering linear retry policy , it would take the client ((1 + 500) * 500 * 2)/2 = 250500 seconds ~= 70 hours to stop retrying. 

```bash
$ ozone fs -ls ofs://ozone2/
23/02/28 07:02:35 WARN ha.OMProxyInfo: OzoneManager address ozone2:9862 for serviceID null remains unresolved for node ID null Check your ozone-site.xml file to ensure ozone manager addresses are configured properly.
23/02/28 07:02:38 INFO retry.RetryInvocationHandler: com.google.protobuf.ServiceException: java.net.UnknownHostException: Invalid host name: local host is: "ozone"; destination host is: "ozone2":9862; java.net.UnknownHostException: Invalid host name: local host is: "ozone"; destination host is: "ozone2":9862; java.net.UnknownHostException; For more details see:  http://wiki.apache.org/hadoop/UnknownHost; For more details see:  http://wiki.apache.org/hadoop/UnknownHost, while invoking $Proxy11.submitRequest over nodeId=null,nodeAddress=ozone2:9862 after 1 failover attempts. Trying to failover after sleeping for 4000ms.
23/02/28 07:02:42 INFO retry.RetryInvocationHandler: com.google.protobuf.ServiceException: java.net.UnknownHostException: Invalid host name: local host is: "ozone"; destination host is: "ozone2":9862; java.net.UnknownHostException: Invalid host name: local host is: "ozone"; destination host is: "ozone2":9862; java.net.UnknownHostException; For more details see:  http://wiki.apache.org/hadoop/UnknownHost; For more details see:  http://wiki.apache.org/hadoop/UnknownHost, while invoking $Proxy11.submitRequest over nodeId=null,nodeAddress=ozone2:9862 after 2 failover attempts. Trying to failover after sleeping for 6000ms.
23/02/28 07:02:48 INFO retry.RetryInvocationHandler: com.google.protobuf.ServiceException: java.net.UnknownHostException: Invalid host name: local host is: "ozone"; destination host is: "ozone2":9862; java.net.UnknownHostException: Invalid host name: local host is: "ozone"; destination host is: "ozone2":9862; java.net.UnknownHostException; For more details see:  http://wiki.apache.org/hadoop/UnknownHost; For more details see:  http://wiki.apache.org/hadoop/UnknownHost, while invoking $Proxy11.submitRequest over nodeId=null,nodeAddress=ozone2:9862 after 3 failover attempts. Trying to failover after sleeping for 8000ms.
```
Adding a check here which tries to resolve the hostname and if it's not successful then fail the op and do not retry.


## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-8041
## How was this patch tested?
Unit tests, manually tried.

After this PR
```bash
$ ozone fs -ls ofs://ozone2/
-ls: Cannot resolve OM host ozone2 in the URI

adoroszlai reviewed (commented):
@sadanand48 I have multiple concerns about the proposed fix.

1. `IllegalArgumentException` is presented to the user with usage help, as if there was a problem in the command syntax.
    ```
    $ ozone fs -ls ofs://om/
    -ls: Cannot resolve OM host om in the URI
    Usage: ozone fs [generic options]
    ...
    ```
2. It does not work for IP addresses (client still retries connection to unreachable IP address for a long time).
3. It does not work for OM HA misconfigured at the client side.
4. It is specific to Ozone FS, does not apply to other Ozone commands.
5. It adds an extra step for Ozone client creation for an edge case.

I would like to suggest a wider discussion of the problem and possible solutions.

sadanand48 commented:
Thanks @adoroszlai for the feedback. Some questions/clarifications:

1. Throwing IllegalArgumentException always prints the command help section, should we change to throw any other Exception so help message won't print like OMException("Cannot connect to OM")?
2. Added a check to handle IP address in the latest commit , it will try to reach the IP and if unreachable throw the error.
3.  I think client should be responsible for configuring right OM address in the config and also currently a warning is thrown if any address cannot be resolved to inform client about incorrect configuration. One idea is to check if atleast one of the OM's defined in the config is reachable by doing a similar check and if not fail.
``` javaif (rpcAddr.isUnresolved()) {
      LOG.warn("OzoneManager address {} for serviceID {} remains unresolved " +
              "for node ID {} Check your ozone-site.xml file to ensure ozone " +
              "manager addresses are configured properly.",
          rpcAddress, serviceId, nodeId);```
4. Added to RpcClient code now, so this is taken care of.
5. True.

kerneltime commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConfigKeys.java:
```diff
@@ -442,7 +442,7 @@ public final class OzoneConfigKeys {
       "ozone.client.failover.max.attempts";
   public static final int OZONE_CLIENT_FAILOVER_MAX_ATTEMPTS_DEFAULT =
-      500;
+      45;
```
Instead of changing this for the client across the board, the CLI should change this. There might be other failover scenarios for long-running jobs where this high count is fine.

sadanand48 replied:
Done.

sumitagrawl commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java:
```diff
@@ -790,4 +791,28 @@ public static String format(List<ServiceInfo> nodes, int port,
+            "OM host " + omHost + " unreachable in the URI");
+      }
+    } catch (IOException e) {
+      throw new RuntimeException(e);
```
We should not encapsulate IOException to RuntimeException, caller is configured to throw IOException. Other issue in communication, we can have error log also

sadanand48 replied:
Done.

sumitagrawl reviewed (changes requested):
@sadanand48 Thanks for working over this, having minor comment, plz check

adoroszlai commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java:
```diff
@@ -790,4 +791,30 @@ public static String format(List<ServiceInfo> nodes, int port,
+   */
+  public static void resolveOmHost(String omHost, int omPort)
+      throws IOException {
+    InetSocketAddress omHostAddress = NetUtils.createSocketAddr(omHost, omPort);
+    if (omHostAddress.isUnresolved()) {
+      throw new IllegalArgumentException(
+          "Cannot resolve OM host " + omHost + " in the URI",
+          new UnknownHostException());
+    }
+    try {
+      if (!omHostAddress.getAddress().isReachable(5000)) {
+        throw new IllegalArgumentException(
+            "OM host " + omHost + " unreachable in the URI");
+      }
+    } catch (IOException e) {
+      LOG.error("Failure in resolving OM host address", e);
+      throw e;
+    }
+  }
```
> Throwing IllegalArgumentException always prints the command help section, should we change to throw any other Exception so help message won't print like OMException("Cannot connect to OM")?

Yes, please.  I think throwing `IOException` in all cases of OM host resolution problems would be better (though I cannot try it currently).

sadanand48 replied:
Done, tried it and it works.
```bash
bash-4.2$ ozone fs -ls ofs://om/
Found 1 items
drwxrwxrwx   - hadoop hadoop          0 2023-03-13 20:13 ofs://om/s3v
bash-4.2$ ozone fs -ls ofs://om1/
ls: Cannot resolve OM host om1 in the URI
bash-4.2$ exit

adoroszlai commented:
Thanks @sadanand48 for updating the patch, @kerneltime, @sumitagrawl for the review.
